File: python/pwr.py
import time

# Function to generate combinations using recursive backtracking
def generate_combinations(char_set, max_length, password, current_string, guesses):
    if current_string == password:
        elapsed_time = time.time() - generate_combinations.start_time
        print(f"Found {password} with {guesses} guesses in {elapsed_time:.6f} seconds")
        exit(0)

    for c in char_set:
        current_string += c
        guesses += 1

        if len(current_string) <= max_length:
            generate_combinations(char_set, max_length, password, current_string, guesses)

        current_string = current_string[:-1]

def main():
    char_set = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890!@#$%^&*()-_=+[]{}|,'.<>?~"

    password = input("Password: ")
    max_length = int(input("Length: "))

    current_string = ""
    guesses = 0

    generate_combinations.start_time = time.time()
    generate_combinations(char_set, max_length, password, current_string, guesses)

if __name__ == "__main__":
    main()

File: c++/misc/pwstor/pwr.cpp
// 6.3 Million Guesses Per Second
// Aid! 0.949004s in 5.7 Million
#include <iostream>
#include <string>
#include <vector>
#include <chrono>

using namespace std;

// Function to generate combinations using recursive backtracking
void generateCombinations(const string& charSet, int maxLength, const string& password, string& currentString, int& guesses) {
    static auto start = chrono::steady_clock::now();

    if (currentString == password) {
        auto end = chrono::steady_clock::now();
        chrono::duration<double> elapsed_seconds = end - start;
        cout << "Found " << password << " with " << guesses << " guesses" << " in " << elapsed_seconds.count() << " seconds" << endl;
        exit(0);
    }

    //cout << currentString << "\n";

    for (char c : charSet) {
        currentString.push_back(c);
        ++guesses;

        if (currentString.length() <= maxLength) {
            generateCombinations(charSet, maxLength, password, currentString, guesses);
        }

        currentString.pop_back();
    }
}

int main() {
    string charSet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890!@#$%^&*()-_=+[]{}|,'.<>?~";

    string password;
    int maxLength;

    cout << "Password: ";
    cin >> password;

    cout << "Length: ";
    cin >> maxLength;

    string currentString;
    int guesses = 0;

    generateCombinations(charSet, maxLength, password, currentString, guesses);

    return 0;
}
